components/encryption/src/manager/mod.rs

```diff
@@ -98,75 +108,86 @@ impl Dicts {
         }
     }
 
-    fn save_key_dict(&mut self, master_key: &dyn Backend) -> Result<()> {
+    fn save_key_dict(&self, master_key: &dyn Backend, key_dict: &mut KeyDictionary) -> Result<()> {
         let file = EncryptedFile::new(&self.base, KEY_DICT_NAME);
         if !master_key.is_secure() {
-            for value in self.key_dict.keys.values_mut() {
+            for value in key_dict.keys.values_mut() {
                 value.was_exposed = true
             }
         }
-        let key_bytes = self.key_dict.write_to_bytes()?;
+        key_dict.current_key_id = self.current_key_id.load(Ordering::Relaxed);
+        let key_bytes = key_dict.write_to_bytes()?;
+        let _lock = self.key_lock.lock().unwrap();
         file.write(&key_bytes, master_key)?;
 
         ENCRYPTION_FILE_SIZE_GAUGE
             .with_label_values(&["key_dictionary"])
             .set(key_bytes.len() as _);
-        ENCRYPTION_DATA_KEY_GAUGE.set(self.key_dict.keys.len() as _);
+        ENCRYPTION_DATA_KEY_GAUGE.set(key_dict.keys.len() as _);
 
         Ok(())
     }
 
     fn save_file_dict(&self) -> Result<()> {
+        let file_dict = self.file_dict.lock().unwrap();
         let file = EncryptedFile::new(&self.base, FILE_DICT_NAME);
-        let file_bytes = self.file_dict.write_to_bytes()?;
+        let file_bytes = file_dict.write_to_bytes()?;
         // File dict is saved in plaintext.
+        let _lock = self.file_lock.lock().unwrap();
```

Comment:
here we are still doing IO while taking `file_dict` lock. The ordering of things should be the following:

for `new_file` etc:
1. take `file_lock`
2. take `file_dict` lock. make a clone of the content. release `file_dict` lock.
3. modify the clone. save it to file
4. take `file_dict` lock and replace the content with the modified clone
5. release `file_lock`

The make the clone on step 2 and 4 more light weight (i.e. not cloning while blocking other `file_dict` reads), we could consider make them an Arc. e.g. `file_dict: Mutex<Arc<FileDictionary>>`, and then:
```
// for read
let file_dict_copy: FileDictionary = {
    let file_dict: Arc<FileDictionary> = self.file_dict.lock().unwrap().clone();
    file_dict.as_ref().clone() // we are cloning the inner file dictionary, not the Arc
};

// ... make modification to file_dict_copy and save file dict file.

// for update
*file_dict.lock().unwrap() = Arc::new(file_dict_copy);
```

I hope my usage of `Mutex<Arc<T>>` is correct, but feel free to point out my mistake. We should do the same for `key_dict`.

Reply:
we need to take `file_lock` before reading from `file_dict`, as we want to avoid a concurrent writer to read stale date while we are doing the update.